omnigibson/systems/macro_particle_system.py

```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
         return state_dict, idx + idx_super
 
 
-VisualParticleSystem.create(
+class MacroPhysicalParticleSystem(PhysicalParticleSystem, MacroParticleSystem):
+    """
+    Particle system class that procedurally generates individual particles that are subject to physics
+    """
+    # Physics rigid body view for keeping track of all particles' state
+    particles_view = None
+
+    # Approximate radius of the macro particle, and distance from particle frame to approximate center
+    _particle_radius = None
+    _particle_offset = None
+
+    @classmethod
+    def initialize(cls):
+        # Run super method first
+        super().initialize()
+
+        # Create the particles head prim -- this is merely a scope prim
+        og.sim.stage.DefinePrim(f"{cls.prim_path}/particles", "Scope")
+
+        # A new view needs to be created every time once sim is playing, so we add a callback now
+        og.sim.add_callback_on_play(name=f"{cls.name}_particles_view", callback=cls._refresh_particles_view)
+
+        # If sim is already playing, refresh particles immediately
+        if og.sim.is_playing():
+            cls._refresh_particles_view()
+
+    @classmethod
+    def _load_new_particle(cls, prim_path, name):
+        # We copy the template prim and generate the new object if the prim doesn't already exist, otherwise we
+        # reference the pre-existing one
+        if not get_prim_at_path(prim_path):
+            omni.kit.commands.execute(
+                "CopyPrim",
+                path_from=cls._particle_object.prim_path,
+                path_to=prim_path,
+            )
+            # Apply RigidBodyAPI to it so it is subject to physics
+            prim = get_prim_at_path(prim_path)
+            UsdPhysics.RigidBodyAPI.Apply(prim)
+            # PhysxSchema.PhysxRigidBodyAPI.Apply(prim)
+        return CollisionVisualGeomPrim(prim_path=prim_path, name=name)
+
+    @classmethod
+    def set_particle_template_object(cls, obj):
+        # Run super method
+        super().set_particle_template_object(obj=obj)
+
+        # Make sure min and max scale are identical
+        assert np.all(cls.min_scale == cls.max_scale), \
+            "Min and max scale should be identical for MacroPhysicalParticleSystem!"
```

Comment:
hmm interesting, would it be more realistic if we actually allow different scales?